File: app/src/main/java/com/simon/harmonichackernews/StoryType.java
package com.simon.harmonichackernews;

import android.content.Context;
import android.content.res.Resources;
import android.text.TextUtils;

import com.simon.harmonichackernews.utils.SettingsUtils;
import com.simon.harmonichackernews.utils.Utils;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Set;

public enum StoryType {
    TOP_STORIES("Top Stories"),
    LAST_24_HOURS("Last 24 hours"),
    LAST_48_HOURS("Last 48 hours"),
    LAST_WEEK("Last week"),
    NEW_STORIES("New Stories"),
    BEST_STORIES("Best Stories"),
    ASK_HN("Ask HN"),
    SHOW_HN("Show HN"),
    HN_JOBS("HN Jobs"),
    CLASSIC(SettingsUtils.FRONT_PAGE_CLASSIC, "classic", true, false, false),
    BEST_COMMENTS(SettingsUtils.FRONT_PAGE_BEST_COMMENTS, "bestcomments", true, true, false),
    HIGHLIGHTS(SettingsUtils.FRONT_PAGE_HIGHLIGHTS, "highlights", true, true, false),
    ACTIVE(SettingsUtils.FRONT_PAGE_ACTIVE, "active", true, false, false),
    FRONT(SettingsUtils.FRONT_PAGE_FRONT, "front", true, false, false),
    BOOKMARKS("Bookmarks"),
    FAVORITES(SettingsUtils.FAVORITES_LABEL),
    UPVOTED(SettingsUtils.UPVOTED_LABEL),
    HISTORY("History"),
    UNKNOWN("");

    private final String label;
    private final String hackerNewsPath;
    private final boolean additionalFrontpage;
    private final boolean commentRows;
    private final boolean frontpageLinkList;

    StoryType(String label) {
        this(label, null, false, false, false);
    }

    StoryType(String label,
              String hackerNewsPath,
              boolean additionalFrontpage,
              boolean commentRows,
              boolean frontpageLinkList) {
        this.label = label;
        this.hackerNewsPath = hackerNewsPath;
        this.additionalFrontpage = additionalFrontpage;
        this.commentRows = commentRows;
        this.frontpageLinkList = frontpageLinkList;
    }

    static ArrayList<CharSequence> buildAdapterLabels(Resources resources, Context ctx, boolean showUserItemLists) {
        String[] sortingOptions = resources.getStringArray(R.array.sorting_options);
        ArrayList<CharSequence> labels = new ArrayList<>(Arrays.asList(sortingOptions));
        int additionalFrontpageIndex = getLabelIndex(labels, BOOKMARKS.label);
        if (additionalFrontpageIndex < 0) {
            additionalFrontpageIndex = Math.min(SettingsUtils.getJobsIndex(resources) + 1, labels.size());
        }
        for (StoryType type : getAdditionalFrontpages()) {
            if (type.isEnabledAdditionalFrontpage(ctx)) {
                labels.add(additionalFrontpageIndex, type.label);
                additionalFrontpageIndex++;
            }
        }
        if (showUserItemLists) {
            int bookmarksIndex = getLabelIndex(labels, BOOKMARKS.label);
            int favoritesIndex = bookmarksIndex >= 0 ? bookmarksIndex + 1 : Math.min(SettingsUtils.getBookmarksIndex(resources) + 1, labels.size());
            labels.add(favoritesIndex, SettingsUtils.FAVORITES_LABEL);
            labels.add(SettingsUtils.UPVOTED_LABEL);
        }
        return labels;
    }

    public static ArrayList<CharSequence> buildStartingPageLabels(Resources resources, Context ctx) {
        return buildStartingPageLabels(resources, SettingsUtils.getEnabledAdditionalFrontpages(ctx));
    }

    public static ArrayList<CharSequence> buildStartingPageLabels(Resources resources, Set<String> enabledFrontpages) {
        String[] startingPageOptions = resources.getStringArray(R.array.starting_page_options);
        ArrayList<CharSequence> labels = new ArrayList<>(Arrays.asList(startingPageOptions));
        for (StoryType type : getAdditionalFrontpages()) {
            if (enabledFrontpages != null && enabledFrontpages.contains(type.label)) {
                labels.add(type.label);
            }
        }
        return labels;
    }

    private static StoryType[] getAdditionalFrontpages() {
        return new StoryType[] {CLASSIC, BEST_COMMENTS, HIGHLIGHTS, ACTIVE, FRONT};
    }

    private static int getLabelIndex(ArrayList<CharSequence> labels, String label) {
        for (int i = 0; i < labels.size(); i++) {
            if (TextUtils.equals(labels.get(i), label)) {
                return i;
            }
        }
        return -1;
    }

    public static StoryType fromLabel(CharSequence label) {
        if (label == null) {
            return UNKNOWN;
        }

        for (StoryType type : values()) {
            if (TextUtils.equals(type.label, label)) {
                return type;
            }
        }
        return UNKNOWN;
    }

    public boolean isAlgolia() {
        return this == LAST_24_HOURS || this == LAST_48_HOURS || this == LAST_WEEK;
    }

    boolean isActive() {
        return this == ACTIVE;
    }

    boolean isFront() {
        return this == FRONT;
    }

    boolean isBookmarks() {
        return this == BOOKMARKS;
    }

    boolean isHistory() {
        return this == HISTORY;
    }

    boolean isFavorites() {
        return this == FAVORITES;
    }

    boolean isUpvoted() {
        return this == UPVOTED;
    }

    boolean isUserItemList() {
        return isFavorites() || isUpvoted();
    }

    boolean usesSavedItemFilter() {
        return isBookmarks() || isUserItemList();
    }

    boolean usesCommentRows() {
        return isBookmarks() || isUserItemList() || commentRows;
    }

    boolean isScrapedFrontpage() {
        return additionalFrontpage && !frontpageLinkList;
    }

    boolean isFrontpageLinkList() {
        return frontpageLinkList;
    }

    public String getHackerNewsPath() {
        return hackerNewsPath;
    }

    String getLabel() {
        return label;
    }

    private boolean isEnabledAdditionalFrontpage(Context ctx) {
        return !additionalFrontpage || (ctx != null && SettingsUtils.isAdditionalFrontpageEnabled(ctx, label));
    }

    String getHackerNewsUrl() {
        switch (this) {
            case TOP_STORIES:
                return Utils.URL_TOP;
            case NEW_STORIES:
                return Utils.URL_NEW;
            case BEST_STORIES:
                return Utils.URL_BEST;
            case ASK_HN:
                return Utils.URL_ASK;
            case SHOW_HN:
                return Utils.URL_SHOW;
            case HN_JOBS:
                return Utils.URL_JOBS;
            default:
                return null;
        }
    }
}

File: app/src/main/java/com/simon/harmonichackernews/StoryType.kt
package com.simon.harmonichackernews

import android.content.Context
import android.content.res.Resources
import com.simon.harmonichackernews.utils.SettingsUtils
import com.simon.harmonichackernews.utils.Utils
import kotlin.math.min

enum class StoryType(
    val label: String,
    val hackerNewsPath: String? = null,
    private val additionalFrontpage: Boolean = false,
    private val commentRows: Boolean = false,
    val isFrontpageLinkList: Boolean = false
) {
    TOP_STORIES("Top Stories"),
    LAST_24_HOURS("Last 24 hours"),
    LAST_48_HOURS("Last 48 hours"),
    LAST_WEEK("Last week"),
    NEW_STORIES("New Stories"),
    BEST_STORIES("Best Stories"),
    ASK_HN("Ask HN"),
    SHOW_HN("Show HN"),
    HN_JOBS("HN Jobs"),
    CLASSIC(SettingsUtils.FRONT_PAGE_CLASSIC, "classic", true, false, false),
    BEST_COMMENTS(SettingsUtils.FRONT_PAGE_BEST_COMMENTS, "bestcomments", true, true, false),
    HIGHLIGHTS(SettingsUtils.FRONT_PAGE_HIGHLIGHTS, "highlights", true, true, false),
    ACTIVE(SettingsUtils.FRONT_PAGE_ACTIVE, "active", true, false, false),
    FRONT(SettingsUtils.FRONT_PAGE_FRONT, "front", true, false, false),
    BOOKMARKS("Bookmarks"),
    FAVORITES(SettingsUtils.FAVORITES_LABEL),
    UPVOTED(SettingsUtils.UPVOTED_LABEL),
    HISTORY("History"),
    UNKNOWN("");

    val isAlgolia: Boolean
        get() = this == LAST_24_HOURS || this == LAST_48_HOURS || this == LAST_WEEK

    val isActive: Boolean
        get() = this == ACTIVE

    val isFront: Boolean
        get() = this == FRONT

    val isBookmarks: Boolean
        get() = this == BOOKMARKS

    val isHistory: Boolean
        get() = this == HISTORY

    val isFavorites: Boolean
        get() = this == FAVORITES

    val isUpvoted: Boolean
        get() = this == UPVOTED

    val isUserItemList: Boolean
        get() = this.isFavorites || this.isUpvoted

    fun usesSavedItemFilter(): Boolean {
        return this.isBookmarks || this.isUserItemList
    }

    fun usesCommentRows(): Boolean {
        return this.isBookmarks || this.isUserItemList || commentRows
    }

    val isScrapedFrontpage: Boolean
        get() = additionalFrontpage && !this.isFrontpageLinkList

    private fun isEnabledAdditionalFrontpage(ctx: Context?): Boolean {
        return !additionalFrontpage || (ctx != null && SettingsUtils.isAdditionalFrontpageEnabled(
            ctx,
            label
        ))
    }

    val hackerNewsUrl: String?
        get() = when (this) {
            TOP_STORIES -> Utils.URL_TOP
            NEW_STORIES -> Utils.URL_NEW
            BEST_STORIES -> Utils.URL_BEST
            ASK_HN -> Utils.URL_ASK
            SHOW_HN -> Utils.URL_SHOW
            HN_JOBS -> Utils.URL_JOBS
            else -> null
        }

    companion object {
        fun buildAdapterLabels(
            resources: Resources,
            ctx: Context?,
            showUserItemLists: Boolean
        ): ArrayList<CharSequence> {
            val sortingOptions = resources.getStringArray(R.array.sorting_options)
            val labels = sortingOptions.mapTo(ArrayList<CharSequence>(sortingOptions.size)) { it }
            var additionalFrontpageIndex = getLabelIndex(labels, BOOKMARKS.label)
            if (additionalFrontpageIndex < 0) {
                additionalFrontpageIndex =
                    min(SettingsUtils.getJobsIndex(resources) + 1, labels.size)
            }
            for (type in additionalFrontpages) {
                if (type.isEnabledAdditionalFrontpage(ctx)) {
                    labels.add(additionalFrontpageIndex, type.label)
                    additionalFrontpageIndex++
                }
            }
            if (showUserItemLists) {
                val bookmarksIndex = getLabelIndex(labels, BOOKMARKS.label)
                val favoritesIndex = if (bookmarksIndex >= 0) bookmarksIndex + 1 else min(
                    SettingsUtils.getBookmarksIndex(resources) + 1, labels.size
                )
                labels.add(favoritesIndex, SettingsUtils.FAVORITES_LABEL)
                labels.add(SettingsUtils.UPVOTED_LABEL)
            }
            return labels
        }

        fun buildStartingPageLabels(resources: Resources, ctx: Context): ArrayList<CharSequence> {
            return buildStartingPageLabels(
                resources,
                SettingsUtils.getEnabledAdditionalFrontpages(ctx)
            )
        }

        fun buildStartingPageLabels(
            resources: Resources,
            enabledFrontpages: Set<String>?
        ): ArrayList<CharSequence> {
            val startingPageOptions = resources.getStringArray(R.array.starting_page_options)
            val labels = startingPageOptions.mapTo(
                ArrayList<CharSequence>(startingPageOptions.size)
            ) { it }
            for (type in additionalFrontpages) {
                if (type.label in enabledFrontpages.orEmpty()) {
                    labels.add(type.label)
                }
            }
            return labels
        }

        private val additionalFrontpages = listOf(
            CLASSIC,
            BEST_COMMENTS,
            HIGHLIGHTS,
            ACTIVE,
            FRONT
        )

        private fun getLabelIndex(labels: List<CharSequence>, label: String): Int =
            labels.indexOfFirst { it.contentEquals(label) }

        fun fromLabel(label: CharSequence?): StoryType {
            return label?.let { value ->
                entries.firstOrNull { type -> type.label.contentEquals(value) }
            } ?: UNKNOWN
        }
    }
}
